Tensor roll op implementation

kobejean opened the pull request:
Closes #10761
Added a tf.manip.roll op that works similarly to numpy's np.roll. This was a feature requested in #10761 and was marked as contributions welcome.

### Usage:

Rolls the elements of a tensor by the offsets of `shift` along the dimensions
of `axis`. Elements that roll passed the last position will wrap around
to the first.
For example:
```
# 't' is [0, 1, 2, 3, 4]
roll(t, shift=2, axis=0) ==> [3, 4, 0, 1, 2]
# shifting along multiple dimensions
# 't' is [[0, 1, 2, 3, 4], [5, 6, 7, 8, 9]]
roll(t, shift=[1, -2], axis=[0, 1]) ==> [[7, 8, 9, 5, 6], [2, 3, 4, 0, 1]]
# shifting along the same axis multiple times
# 't' is [[0, 1, 2, 3, 4], [5, 6, 7, 8, 9]]
roll(t, shift=[2, -3], axis=[1, 1]) ==> [[1, 2, 3, 4, 0], [6, 7, 8, 9, 5]]
```
shift: `shift[i]` specifies the number of places by which elements are shifted
  along the dimension specified by `axis[i]`. Negative shifts will roll the
  elements in the opposite direction.
axis: `axis[i]` specifies the dimension that the shift `shift[i]` should occur.
  if the same axis is referenced more than once, the total shift for that axis
  will be the sum of all the shifts that belong to that axis.
output: Has the same shape and size as the input. The elements are shifted by
  the offsets of `shift` along the dimensions of `axis`.

### Unit tests:
```
[==========] Running 24 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 24 tests from RollOpTest
[ RUN      ] RollOpTest.ScalarIndices
[       OK ] RollOpTest.ScalarIndices (5 ms)
[ RUN      ] RollOpTest.ScalarIndices_NoMemcpy
[       OK ] RollOpTest.ScalarIndices_NoMemcpy (0 ms)
[ RUN      ] RollOpTest.ScalarIndices_Complex
[       OK ] RollOpTest.ScalarIndices_Complex (0 ms)
[ RUN      ] RollOpTest.Simple_TwoD32
[       OK ] RollOpTest.Simple_TwoD32 (0 ms)
[ RUN      ] RollOpTest.Simple_TwoD32_NoMemcpy
[       OK ] RollOpTest.Simple_TwoD32_NoMemcpy (0 ms)
[ RUN      ] RollOpTest.Simple_ThreeD32
[       OK ] RollOpTest.Simple_ThreeD32 (1 ms)
[ RUN      ] RollOpTest.Simple_ThreeD32_NoMemcpy
[       OK ] RollOpTest.Simple_ThreeD32_NoMemcpy (0 ms)
[ RUN      ] RollOpTest.Simple_TwoD64
[       OK ] RollOpTest.Simple_TwoD64 (0 ms)
[ RUN      ] RollOpTest.Simple_TwoD64_NoMemcpy
[       OK ] RollOpTest.Simple_TwoD64_NoMemcpy (0 ms)
[ RUN      ] RollOpTest.Simple_ThreeD64
[       OK ] RollOpTest.Simple_ThreeD64 (0 ms)
[ RUN      ] RollOpTest.Simple_ThreeD64_NoMemcpy
[       OK ] RollOpTest.Simple_ThreeD64_NoMemcpy (0 ms)
[ RUN      ] RollOpTest.ZeroShift_ThreeD32
[       OK ] RollOpTest.ZeroShift_ThreeD32 (0 ms)
[ RUN      ] RollOpTest.ZeroShift_ThreeD32_NoMemcpy
[       OK ] RollOpTest.ZeroShift_ThreeD32_NoMemcpy (0 ms)
[ RUN      ] RollOpTest.ZeroSize_ThreeD32
[       OK ] RollOpTest.ZeroSize_ThreeD32 (0 ms)
[ RUN      ] RollOpTest.ZeroSize_ThreeD32_NoMemcpy
[       OK ] RollOpTest.ZeroSize_ThreeD32_NoMemcpy (0 ms)
[ RUN      ] RollOpTest.OneSize_ThreeD32
[       OK ] RollOpTest.OneSize_ThreeD32 (1 ms)
[ RUN      ] RollOpTest.OneSize_ThreeD32_NoMemcpy
[       OK ] RollOpTest.OneSize_ThreeD32_NoMemcpy (0 ms)
[ RUN      ] RollOpTest.DuplicateShifts_TwoD32
[       OK ] RollOpTest.DuplicateShifts_TwoD32 (0 ms)
[ RUN      ] RollOpTest.DuplicateShifts_TwoD32_NoMemcpy
[       OK ] RollOpTest.DuplicateShifts_TwoD32_NoMemcpy (0 ms)
[ RUN      ] RollOpTest.Error_InputMustBeVectorOrHigher
[       OK ] RollOpTest.Error_InputMustBeVectorOrHigher (0 ms)
[ RUN      ] RollOpTest.Error_AxisMustBeScalarOrVector
[       OK ] RollOpTest.Error_AxisMustBeScalarOrVector (0 ms)
[ RUN      ] RollOpTest.Error_ShiftMustBeScalarOrVector
[       OK ] RollOpTest.Error_ShiftMustBeScalarOrVector (0 ms)
[ RUN      ] RollOpTest.Error_ShiftAndAxisMustBeSameSize
[       OK ] RollOpTest.Error_ShiftAndAxisMustBeSameSize (0 ms)
[ RUN      ] RollOpTest.Error_AxisOutOfRange
[       OK ] RollOpTest.Error_AxisOutOfRange (0 ms)
[----------] 24 tests from RollOpTest (7 ms total)

[----------] Global test environment tear-down
[==========] 24 tests from 1 test case ran. (8 ms total)
[  PASSED  ] 24 tests.
```

```
//tensorflow/python/kernel_tests:manip_ops_test                          PASSED in 0.9s
```

### Benchmarks:
```
Running main() from test_main.cc
Benchmark                     Time(ns) Iterations
-------------------------------------------------
BM_cpu_roll_outer/256/256        46739      16065	 5608.7MB/s 1402.2M items/s
BM_cpu_roll_outer/512/512       124247       5135	 8439.5MB/s 2109.9M items/s
BM_cpu_roll_outer/1024/1024     638245       1149	 6571.6MB/s 1642.9M items/s
BM_cpu_roll_outer/2048/2048    5261260        100	 3188.8MB/s 797.2M items/s
BM_cpu_roll_all/256/256         135631       5353	 1932.8MB/s 483.2M items/s
BM_cpu_roll_all/512/512         286927       2263	 3654.5MB/s 913.6M items/s
BM_cpu_roll_all/1024/1024       981417        727	 4273.7MB/s 1068.4M items/s
BM_cpu_roll_all/2048/2048      4549074        149	 3688.1MB/s 922.0M items/s
```

I had made a pull request for this before but accidentally closed it

kobejean committed:
Rebase for roll op code

kobejean committed:
Roll op: Supports CPU thread pooling

fix namespace error

kobejean committed:
Roll op: Optimization

kobejean committed:
Roll op: Pylint fix

kobejean committed:
Roll op: Updated documentation

kobejean committed:
Roll op: Two versions for CPU thread pooling

kobejean committed:
Roll op: Huge CPU speed up

Fixed thread pooling issue that was due to a bad cost_per_unit parameter

Also improved readability

kobejean committed:
Roll op: Rough draft of DoRollV2

DoRollV2 copies memory in groups instead of element by element.
Not thoroughly tested yet.

Polished DoRollV2 algorithm

kobejean committed:
Roll op: Restrict tensor size for GPU implementation

kobejean committed:
Roll op: Fixed clang-format and missing include

kobejean committed:
Roll op: Minor change

kobejean committed:
Roll op GPU bug fix

Roll op GPU bug fix

GPU bug fix

Roll op GPU bug fix

Roll op GPU fix

Roll GPU test BUILD update

kobejean committed:
Roll op: Remove GPU code

Fully remove roll op GPU code

Remove compile_cpu.sh

kobejean committed:
Roll op: polymorphism & GPU attempt

kobejean committed:
Roll op: Added support for gradients

Added compile script

kobejean committed:
Roll op: Migrated to manip namespace

kobejean committed:
Remove roll from user_ops

kobejean committed:
Roll op: Fixes problems with array_ops_test.py and a size 1 dimension bug

kobejean committed:
Half migrated to manip

Half migrated to tf.manip

kobejean committed:
Roll op: Migrated to manip

Migrated to tf.manip

Roll op registered

Roll op uses InlinedVector

Small improvements

kobejean committed:
Roll op: Revert array op changes

kobejean committed:
Roll op: Api def fix

kobejean committed:
Merge branch 'roll'

tensorflow-jenkins commented:
Can one of the admins verify this patch?

kobejean committed:
Merge branch 'master' into master

jhseu reviewed (commented):
@yzhwang , you might be a better reviewer since you've been optimizing transpose. Mind taking a look?

Just some basic comments.

yzhwang commented on tensorflow/core/kernels/roll_op.cc:
```diff
@@ -0,0 +1,306 @@
+    // inner shift dimension (inner most shifted dimension)
+    int64 isd = 0;
+    for (int d = D - 1; d >= 0; d--) {
+      if (!isd && shift_mod_sum[d]) isd = d;
```
Sure! I can take another pass once the current comments have being resolved.

kobejean replied:
Thanks @jhseu I've added those changes. Btw @yzhwang the op is added in a new tf.manip namespace. This was something I was requested to do when I opened a PR here #12786. I accidentally closed that PR when rebasing my fork so I had to open a new one here. Anyway thanks!

kobejean committed:
Roll op: review changes

kobejean commented on tensorflow/core/kernels/roll_op.cc:
```diff
@@ -0,0 +1,306 @@
+    // if there are any duplicate axes, shift_mod_sum will have the
+    // total modulo sum of shifts for each dimension
+    gtl::InlinedVector<int, 4> shift_mod_sum;
+    for (int d = 0; d < D; d++) shift_mod_sum[d] = 0;  // default is 0
```
Thanks for the suggestion. I tried using push_back but I think the problem was that InlinedVector<int, 4> initializes with 4 empty elements so using push_back would add elements in the wrong place. Maybe InlinedVector is not the best choice? I just looked at what other op kernels had and tried to mimick them. I think I got it to initialize with the right number of elements though.

asimshankar commented on tensorflow/core/ops/manip_ops.cc:
```diff
@@ -0,0 +1,62 @@
+    .Attr("Taxis: {int32,int64}")
+    .SetShapeFn(shape_inference::UnchangedShape)
+    .Doc(R"doc(
+Rolls the elements of a tensor by the offsets of `shift` along the dimensions
```
Can you make a one-line summary and then add details in the following paragraph?

(That leads to nicer docstrings in the generated Python code)

asimshankar replied:
Ping?

kobejean replied:
@asimshankar I ended up removing the .Doc line because I noticed that it wasn't being used for other ops like in `array_ops.cc`. The summary and description is defined in `tensorflow/core/api_def/base_api/api_def_Roll.pbtxt` and has the changes you've suggested.

asimshankar reviewed (commented):
(Good for API addition. Thanks!)

kobejean committed:
Roll op: API review changes

Roll op: Docstring fix

yzhwang commented on tensorflow/python/ops/manip_grad.py:
```diff
@@ -0,0 +1,32 @@
+
+@ops.RegisterGradient("Roll")
+def _RollGrad(op, grad):
+  # The gradient should be just the roll reversed
```
The gradient __is__ the reversed roll of the output. Because roll op can be written as matrix multiplication then it can be proved the gradient is the multiplication of the roll op output and the transpose matrix of the original matrix used for roll, which is a reversed roll. You do not need to put too much details here, but you need to be sure that it __is__ the reversed roll of the output. So please change the comment here.

yzhwang commented on tensorflow/core/kernels/roll_op.cc:
```diff
@@ -0,0 +1,306 @@
+                          RollOp<CPUDevice, type, int64, int64>)
+
+TF_CALL_ALL_TYPES(REGISTER_CPU);
+REGISTER_CPU(bfloat16);
```
Why do you have this line? Do you have implementation/test cases for this?

yzhwang reviewed (changes requested):
Thank you for the PR @kobejean ! Finished the first round. Haven't looked at tests yet. Will do it in another round.

kobejean committed:
Roll op: Review changes round 1

kobejean commented:
@yzhwang Thanks for the review! Just added the changes!

kobejean commented on tensorflow/core/kernels/roll_op.cc:
```diff
@@ -0,0 +1,306 @@
+      const int ds = std::max<int>(static_cast<int>(input.dim_size(a)), 1);
+      const int sum = shift_mod_sum[a] + static_cast<int>(shift_flat(m));
+      // modulo that works with negatives: ((x % y) + y) % y
+      shift_mod_sum[a] = (sum % ds + ds) % ds;
```
shift_flat can have negative values if the user wants the elements shifted in the opposite direction.

kobejean committed:
Roll op: resolve conflicts

kobejean committed:
Roll op: Resolve conflicts

kobejean committed:
Merge remote-tracking branch 'upstream/master' into resolve-m

kobejean committed:
Roll op: clang-tidy

rmlarsen commented:
@yzhwang Could you please take another look?

yzhwang commented on tensorflow/core/kernels/roll_op.cc:
```diff
@@ -0,0 +1,334 @@
+/* Copyright 2015 The TensorFlow Authors. All Rights Reserved.
```
2018?

yzhwang commented on tensorflow/core/kernels/roll_op.cc:
```diff
@@ -0,0 +1,334 @@
+            // offset = shift[j] + ... other offsets
+            // offset - dim_range[j] = -threshold[j] + ... other offsets
+            // thus we undo our previous offset as well as add a new offset of
+            // -threshold[j] in one opperation
```
opperation->operation

yzhwang commented on tensorflow/core/kernels/roll_op_test.cc:
```diff
@@ -0,0 +1,481 @@
+/* Copyright 2015 The TensorFlow Authors. All Rights Reserved.
```
2018

yzhwang commented on tensorflow/core/kernels/roll_op_test.cc:
```diff
@@ -0,0 +1,481 @@
+  test::ExpectTensorEqual<string>(expected, *GetOutput(0));
+}
+
+TEST_F(RollOpTest, DuplicateShifts_TwoD32) {
```
Consider switching duplicate to multi since you are testing combining multiple shifts on one axis.

yzhwang commented on tensorflow/core/kernels/roll_op_test.cc:
```diff
@@ -0,0 +1,481 @@
+  EXPECT_TRUE(StringPiece(s.ToString()).contains("is out of range")) << s;
+}
+
+static Graph* RollGraph(const TensorShape& shape, int isd) {
```
What is isd here?

kobejean replied:
I'll add some comments to clarify. This also stands for inner shift dimension. Basically I shift the isd and all outer dimensions. When I benchmark with a 2-D tensor if isd=0 then since dimension 1 is on the inside of dimension 0, dimension 1 will not be shifted and dimension 0 will. If isd=1 then both dimension 1 and 0 get shifted. This allows me to compare performance between just shifting the outer dimension and shifting all dimensions.

yzhwang commented on tensorflow/core/ops/manip_ops.cc:
```diff
@@ -0,0 +1,67 @@
+/* Copyright 2016 The TensorFlow Authors. All Rights Reserved.
```
2018 :-)

yzhwang commented on tensorflow/python/kernel_tests/manip_ops_test.py:
```diff
@@ -0,0 +1,126 @@
+# Copyright 2015 The TensorFlow Authors. All Rights Reserved.
```
2018 :-)

yzhwang reviewed (approved):
Thanks for the change! Both code and tests LGTM, just had some minor nits, all should be easy fix. Also please resolve on conflict in the BUILD file before you merge in the code.

kobejean committed:
Merge remote-tracking branch 'upstream/master'

kobejean committed:
Roll op: Review round 2 changes

Roll op: fixed BUILD file

Roll op: api docs update

kobejean commented:
Thanks for the review! Just added the last changes!

kobejean committed:
Roll op: failure fixes 1

- updates goldens and fixes api compatibility issue
- fixes python op test issue for windows
- fixes makefile issues

kobejean commented:
@rmlarsen I think I fixed the failures. Can we run the checks again?

kobejean committed:
Roll op: Windows CMake failure fix

Windows CMake checks were failing because numpy was on an older version that did not support np.roll on multiple shifts that was use to check the correctness of tf.manip.roll.  manip_ops_test.py now checks for numpy version 1.12.0 before testing multiple shifts, otherwise it'll just test single shift roll.

kobejean commented:
@rmlarsen @yzhwang Ok this should do it. Sorry can we run the checks again?

kobejean committed:
Roll op: pylint changes

rmlarsen commented:
@asimshankar thanks. Sorry I missed that you had already done an API review on this!
